src/crypto/secretstream/secretstream_macros.rs

```diff
@@ -0,0 +1,354 @@
+        let mlen = m.len();
+        if m.len() > messagebytes_max() {
+            return Err(());
+        }
+        let clen = mlen + ABYTES;
+        let mut c = Vec::with_capacity(clen);
+        let (ad_p, ad_len) = ad.map(|ad| (ad.as_ptr(), ad.len()))
+                               .unwrap_or((ptr::null(), 0));
+
+        let rc = unsafe {
+            $push_name(&mut self.0,
+                       c.as_mut_ptr(),
+                       &mut (clen as c_ulonglong),
+                       m.as_ptr(),
+                       mlen as c_ulonglong,
+                       ad_p,
+                       ad_len as c_ulonglong,
+                       tag)
+        };
+        if rc != 0 {
+            return Err(());
+        }
+
+        unsafe { c.set_len(clen) };
+        Ok(c)
+    }
+
+    /// Encrypts a message `m` and the tag `Tag::Message`. Optionally attaches
+    /// unencrypted additional data `ad`. All data is authenticated.
+    pub fn aencrypt_message(&mut self, m: &[u8], ad: Option<&[u8]>) -> Result<Vec<u8>, ()> {
+        self.aencrypt(m, ad, TAG_MESSAGE)
+    }
+
+    /// Encrypts a message `m` and the tag `Tag::Push`. Optionally attaches
+    /// unencrypted additional data `ad`. All data is authenticated.
+    pub fn aencrypt_push(&mut self, m: &[u8], ad: Option<&[u8]>) -> Result<Vec<u8>, ()> {
+        self.aencrypt(m, ad, TAG_PUSH)
+    }
+
+    /// Encrypts a message `m` and the tag `Tag::Rekey`. Rekeys the `Encryptor`,
+    /// updating the internal state. Optionally attaches unencrypted additional
+    /// data `ad`. All data is authenticated.
+    pub fn aencrypt_rekey(&mut self, m: &[u8], ad: Option<&[u8]>) -> Result<Vec<u8>, ()> {
+        self.aencrypt(m, ad, TAG_REKEY)
+    }
+
+    /// Encrypts a message `m` and the tag `Tag::Finalize`. Optionally attaches
+    /// unencrypted additional data `ad`. All data is authenticated. This method
+    /// consumes the `Encryptor`, so it can no longer be used after a call to
+    /// `finalize`.
+    pub fn aencrypt_finalize(mut self, m: &[u8], ad: Option<&[u8]>) -> Result<Vec<u8>, ()> {
+        self.aencrypt(m, ad, TAG_FINAL)
+    }
+
+    /// This method explicitly re-keys the `Encryptor` and updates its state, but
+    /// doesn't add any information about the key change to the stream. If this
+    /// function is used to create an encrypted stream, the decryption process
+    /// must call that function at the exact same stream location.
+    /// See also the method `aencrypt_rekey`.
+    pub fn rekey(&mut self) {
+        unsafe {
+            $rekey_name(&mut self.0);
+        }
+    }
+}
+
+/// `Decryptor` contains the state for multi-part (streaming) computations. This
+/// allows the caller to process encryption of a sequence of multiple messages.
+/// After the last message of a valid stream containing the tag `Tag::Finalized`
+/// is decrypted, the `Decryptor` may no longer verify and decrypt messages or
+/// re-key itself.
+pub struct Decryptor {
```

Comment:
Same name discussion as above. Although I hate the name ‘Opener’, so I’m open to suggestions.